### PR TITLE
feat: CNS-1009 - estimated rewards query by cu

### PR DESCRIPTION
**Pull request by oren-lava**
# Description

Closes: #XXXX

Re-implemented the `estimate-rewards` query to present a more accurate rewards estimation, by CU usage.

---

## Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] read the [contribution guide](https://github.com/lavanet/lava/blob/main/CONTRIBUTING.md)
* [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title, you can find examples of the prefixes below:
    
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] targeted the `main` branch
* [ ] provided a link to the relevant issue or specification
* [ ] reviewed "Files changed" and left comments if necessary
* [ ] included the necessary unit and integration tests
* [ ] updated the relevant documentation or specification, including comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] confirmed all CI checks have passed

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic, API design and naming, documentation is accurate, tests and test coverage

### CODE REVIEW

**Review by omerlavanet — CHANGES_REQUESTED** [review]
missing adding tracked cu in the estimation to the provider in case the delegator adds more stake

File: x/dualstaking/types/delegate.go [hunk 1]
```diff
@@ -34,7 +33,7 @@ func NewDelegation(delegator, provider string, blockTime time.Time, tokenDenom s
 		Delegator: delegator,
 		Provider:  provider,
 		Amount:    sdk.NewCoin(tokenDenom, sdk.ZeroInt()),
-		Timestamp: utils.NextMonth(blockTime).UTC().Unix(),
+		Timestamp: blockTime.AddDate(0, 0, 7).UTC().Unix(),
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/dualstaking/types/delegate.go:36]: done. see https://github.com/lavanet/lava/commit/790cd4c9ea41cd73221f659dd463878f07f3fe3f

**Comment by oren-lava** [comment]
> missing adding tracked cu in the estimation to the provider in case the delegator adds more stake

done. see https://github.com/lavanet/lava/commit/615cbd80e1ec9e14f1e396af89c98318cf6eed08

File: proto/lavanet/lava/subscription/query.proto [hunk 2]
```diff
@@ -140,12 +153,15 @@ message EstimatedRewardInfo {
 }
 
 message QueryEstimatedRewardsResponse {
-  repeated EstimatedRewardInfo info = 1 [(gogoproto.nullable) = true];
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply proto/lavanet/lava/subscription/query.proto:143]: source is equal to "reward source" (which can be IPRPC, subscription or boost) and chain ID

File: x/subscription/keeper/grpc_query_estimated_rewards.go [hunk 3]
```diff
@@ -3,127 +3,356 @@ package keeper
...
-	found = false
-	for _, k := range emissions {
-		if k.ChainID == req.ChainId {
-			specEmission = k
-			found = true
-			break
+	if len(qRes.Rewards) == 0 {
+		return sdk.NewCoins(), nil
+	}
+
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/subscription/keeper/grpc_query_estimated_rewards.go:186]: done. see https://github.com/lavanet/lava/commit/8d597eb11dce31850af77c905877bc644e46bdd5

File: x/subscription/keeper/grpc_query_estimated_rewards.go [hunk 4]
```diff
@@ -3,127 +3,356 @@ package keeper
...
-			found = true
-			entry = e
-		}
-	}
-	if !found {
-		return nil, fmt.Errorf("provider not found in stake entries for chain ID: %s", req.ChainId)
+	details := []utils.Attribute{
+		utils.LogAttr("block", ctx.BlockHeight()),
+		utils.LogAttr("provider", req.Provider),
+		utils.LogAttr("delegator_amount", req.AmountDelegator),
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/subscription/keeper/grpc_query_estimated_rewards.go]: done. see https://github.com/lavanet/lava/commit/8d597eb11dce31850af77c905877bc644e46bdd5

File: x/subscription/keeper/grpc_query_estimated_rewards.go [hunk 5]
```diff
@@ -3,127 +3,356 @@ package keeper
...
 	"google.golang.org/grpc/status"
 )
 
-func (k Keeper) EstimatedRewards(goCtx context.Context, req *types.QueryEstimatedRewardsRequest) (*types.QueryEstimatedRewardsResponse, error) {
+func (k Keeper) EstimatedProviderRewards(goCtx context.Context, req *types.QueryEstimatedProviderRewardsRequest) (*types.QueryEstimatedRewardsResponse, error) {
 	if req == nil {
 		return nil, status.Error(codes.InvalidArgument, "invalid request")
 	}
-
 	ctx := sdk.UnwrapSDKContext(goCtx)
```
[the remark this answers is not in the document]
Comment (oren-lava) [reply x/subscription/keeper/grpc_query_estimated_rewards.go]: there's no way to check that a ctx is a query ctx. We used a copy of the ctx to make sure it doesn't use the original ctx. see here: https://github.com/lavanet/lava/commit/3cd6061ecd25bb833dade00962dffbca712f9d11

[the remark this answers is not in the document]
Comment (oren-lava) [reply x/subscription/keeper/grpc_query_estimated_rewards.go]: see also: https://github.com/lavanet/lava/commit/eba617b43b6d8ef959ec9cc8a529979aba9e66e2

**Review by omerlavanet — APPROVED** [review]

**Pull request merged**